File: CsChatClient/Models/ConversationInfo.cs
using System;
using CsChatClient.Messages;
using Newtonsoft.Json;

namespace CsChatClient.Models
{
    public class ConversationInfo : ISerializable
    {
        public Conversation Conversation { get; set; }
        public MessageEx LastMessage { get; set; }
        public long Timestamp { get; set; }
        public string Draft { get; set; }
        public UnreadCount UnreadCount { get; set; }
        public int IsTop { get; set; }
        public bool IsSilent { get; set; }

        public void Serialize(JsonWriter writer)
        {
            throw new NotImplementedException();
        }

        public bool Unserialize(JsonReader reader)
        {
            Conversation = new Conversation();
            while (reader.Read())
            {
                switch (reader.TokenType)
                {
                    case JsonToken.PropertyName:
                        if (reader.Value.Equals("conversationType"))
                        {
                            Conversation.Type = (ConversationType)JsonTools.GetNextInt(reader);
                        }
                        else if (reader.Value.Equals("target"))
                        {
                            Conversation.Target = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("line"))
                        {
                            Conversation.Line = JsonTools.GetNextInt(reader);
                        }
                        else if (reader.Value.Equals("lastMessage"))
                        {
                            LastMessage = (MessageEx)JsonTools.GetNextObject(reader, false, typeof(MessageEx));
                        }
                        else if (reader.Value.Equals("timestamp"))
                        {
                            Timestamp = JsonTools.GetNextBigInt(reader);
                        }
                        else if (reader.Value.Equals("draft"))
                        {
                            Draft = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("unreadCount"))
                        {
                            UnreadCount = (UnreadCount)JsonTools.GetNextObject(reader, false, typeof(UnreadCount));
                        }
                        else if (reader.Value.Equals("isTop"))
                        {
                            IsTop = JsonTools.GetNextInt(reader);
                        }
                        else if (reader.Value.Equals("isSilent"))
                        {
                            IsSilent = JsonTools.GetNextBoolean(reader);
                        }
                        break;
                    case JsonToken.EndObject:
                        return true;
                }
            }
            return false;
        }
    }
}

File: CppChatClient/model/ConversationInfo.cpp
//
//  ConversationInfo.cpp
//  ChatClient
//
//  Created by Hao Jia on 2020/1/30.
//  Copyright © 2020 WF Chat. All rights reserved.
//

#include "ConversationInfo.h"
#include "../JsonTools.h"
    
namespace WFCLib {

void ConversationInfo::Unserialize(const void *pvalue) {
    const Value& value = *((const Value *)pvalue);
    if (!value.IsObject()) {
        return;
    }

    int conversationType;
    getValue(value, "conversationType", conversationType);
    std::string target;
    getValue(value, "target", target);
    int line;
    getValue(value, "line", line);
    
    conversation.conversationType = (ConversationType)conversationType;
    conversation.target = target;
    conversation.line = line;
    
//            TMessage lastMessage;
    getValue(value, "lastMessage", lastMessage);
//            int64_t timestamp;
    getValue(value, "timestamp", timestamp);
//            std::string draft;
    getValue(value, "draft", draft);
//            TUnreadCount unreadCount;
    getValue(value, "unreadCount", unreadCount);
//            bool isTop;
    getValue(value, "isTop", isTop);
//            bool isSilent;
    getValue(value, "isSilent", isSilent);
}

void ConversationInfo::Serialize(void *pwriter) const {
    Writer<StringBuffer> &writer = *((Writer<StringBuffer> *)pwriter);

    writer.StartObject();

//            int conversationType;
    writer.String("conversationType");
    writer.Int(conversation.conversationType);
//            std::string target;
    writer.String("target");
    writer.String(conversation.target);
//            int line;
    writer.String("line");
    writer.Int(conversation.line);
//            TMessage lastMessage;
    writer.String("lastMessage");
    lastMessage.Serialize(pwriter);
//            int64_t timestamp;
    writer.String("timestamp");
    writer.Int64(timestamp);
//            std::string draft;
    writer.String("draft");
    writer.String(draft);
//            TUnreadCount unreadCount;
    writer.String("unreadCount");
    unreadCount.Serialize(pwriter);
//            bool isTop;
    writer.String("isTop");
    writer.Bool(isTop);
//            bool isSilent;
    writer.String("isSilent");
    writer.Bool(isSilent);

    writer.EndObject();
}

};
